libraries/AC_AutoTune/AC_AutoTune.h

```diff
@@ -215,14 +324,120 @@ class AC_AutoTune {
     } level_problem;
 
+    // parameters
     AP_Int8  axis_bitmask;
     AP_Float aggressiveness;
     AP_Float min_d;
+    AP_Float vel_hold_gain;
 
     // copies of object pointers to make code a bit clearer
-    AC_AttitudeControl_Multi *attitude_control;
+    AC_AttitudeControl *attitude_control;
     AC_PosControl *pos_control;
     AP_AHRS_View *ahrs_view;
     AP_InertialNav *inertial_nav;
     AP_Motors *motors;
+
+    // Functions added for heli autotune
+
+    // Add additional updating gain functions specific to heli
+    // generic method used by subclasses to update gains for the rate ff up tune type
+    virtual void updating_rate_ff_up_all(AxisType test_axis)=0;
+    // generic method used by subclasses to update gains for the rate ff down tune type
+    virtual void updating_rate_ff_down_all(AxisType test_axis)=0;
+    // generic method used by subclasses to update gains for the max gain tune type
+    virtual void updating_max_gains_all(AxisType test_axis)=0;
+
+    // Feedforward test used to determine Rate FF gain
+    void rate_ff_test_init();
+    void rate_ff_test_run(float max_angle_cds, float target_rate_cds, float dir_sign);
+
+    // dwell test used to perform frequency dwells for rate gains
+    void dwell_test_init(float filt_freq);
+    void dwell_test_run(uint8_t freq_resp_input, float start_frq, float stop_frq, float &dwell_gain, float &dwell_phase);
+
+    // dwell test used to perform frequency dwells for angle gains
+    void angle_dwell_test_init(float filt_freq);
+    void angle_dwell_test_run(float start_frq, float stop_frq, float &dwell_gain, float &dwell_phase);
+
+    float waveform(float time, float time_record, float waveform_magnitude, float wMin, float wMax);
+
+    uint8_t  ff_test_phase;                         // phase of feedforward test
+    float    test_command_filt;                     // filtered commanded output
+    float    test_rate_filt;                        // filtered rate output
+    float    command_out;
+    float    test_tgt_rate_filt;                    // filtered target rate
+    float    filt_target_rate;
+    bool     ff_up_first_iter;                      //true on first iteration of ff up testing
+    float    test_gain[20];                             // gain of output to input
+    float    test_freq[20];
+    float    test_phase[20];
+    float    dwell_start_time_ms;
+    uint8_t  freq_cnt;
+    uint8_t  freq_cnt_max;
+    float    curr_test_freq;
+    float    curr_test_gain;
+    float    curr_test_phase;
+    Vector3f start_angles;
+    uint32_t settle_time;
+    uint32_t phase_out_time;
+    float    waveform_freq_rads;  //current frequency for chirp waveform
+    float    start_freq;  //start freq for dwell test
+    float    stop_freq;   //ending freq for dwell test
+    float    trim_pff_out;  // trim output of the PID rate controller for P, I and FF terms
+    float    trim_meas_rate;  // trim measured gyro rate
+
+    //variables from rate FF test
```

Comment:
similar to the dwell test I think we should move the rate FF test related methods and variables to the Heli class.

Reply:
done